### PR TITLE
Add support for RFC8439 variant of ChaCha20

### PR DESCRIPTION
**Pull request by sipa**
Based on and replaces part of #25361, part of the BIP324 project (#27634). See also #19225 for background.

There are two variants of ChaCha20 in use. The currently implemented one uses a 64-bit nonce and a 64-bit block counter, while the one used in RFC8439 (and thus BIP324) uses a 96-bit nonce and 32-bit block counter. This PR changes the logic to use the 96-bit nonce variant, though in a way that's compatible with >256 GiB output (by automatically incrementing the first 32-bit part of the nonce when the block counter overflows).

For those who reviewed the original PR, the biggest change is here that the 96-bit nonce is passed as a Nonce96 type (pair of 32-bit + 64-bit integer) rather than a 12-byte array.

### CODE REVIEW

**Comment by Sjors** [comment]
#19225 :-)

File: src/test/crypto_tests.cpp [hunk 1]
```diff
@@ -184,6 +184,33 @@ static void TestChaCha20(const std::string &hex_message, const std::string &hexk
...
+        c20.SetRFC8439(nonce, counter);
+        std::vector<unsigned char> keystream;
+        auto expected_keystream = ParseHex(hex_expected_keystream);
+        keystream.resize(expected_keystream.size());
+        c20.Keystream(keystream.data(), expected_keystream.size());
+        BOOST_CHECK_EQUAL(HexStr(keystream), hex_expected_keystream);
+    }
+
+    if (!hex_input.empty()) {
+        assert(hex_input.size() == hex_expected_output.size());
```
Comment (theStack) [inline src/test/crypto_tests.cpp:202]: tiny nit (probably only worth changing if you have to retouch): could check this assert unconditionally, to not allow test cases where `hex_input` is empty but `hex_expected_output` is not

**Review by theStack — COMMENTED** [review]
Concept ACK

**Review by theStack — APPROVED** [review]
Code-review ACK 200e1a292b277d6c32dbd6081d38b715cbc079cb

Verified that the implementation matches the RFC8439 specification (https://datatracker.ietf.org/doc/html/rfc8439#section-2.3) and checked that the test vectors match the ones from the linked sources (https://tools.ietf.org/html/draft-agl-tls-chacha20poly1305-04#section-7 and  https://datatracker.ietf.org/doc/html/rfc8439#page-30).

Note for other reviewers (that get as easily confused as me):  we don't store the constants in the `input` array, so our indices are 4 less, e.g. `input[8]` in our code corresponds to `Word 12` in the RFC.

**Comment by achow101** [comment]
Is the current ChaCha20 implementation used for anything that needs backwards compatibility?

**Comment by achow101** [comment]
Assuming it doesn't break a whole lot of tests, I would prefer to change it always use the new version as it removes the need to check that everything is using the right variant.

**Comment by achow101** [comment]
ACK 7f2a985147ef541123c65d5db1c3fc3e533fd4ce

This PR fundamentally doesn't change our implementation of ChaCha20, just its interface so that it matches RFC8439. It otherwise behaves in the same way as previously. The new test vectors match those of the RFC.

**Review by theStack — COMMENTED** [review]
Nice and indeed much simpler now, the actual diff is smaller than I expected. Will re-review tomorrow.

Meanwhile, I thought it would be nice to have a test case triggering the 32-bit block counter overflow to check for >256 GiB output compatibility and wrote one (test data is created with PyCryptodome, see commit message), maybe it's worthwhile to include it here: https://github.com/theStack/bitcoin/commit/61b41ad16e35306485a1db508d624d3ddab0ca0a (the next commit https://github.com/theStack/bitcoin/commit/ff289308bf7af8184cc0ecaafe6f5e1842b8f17b goes one step further and checks if the overflow really happened by inspecting the input state before/after test case execution, but not sure if it's worth it to pollute the ChaCha20 class with test-only methods).

**Review by theStack — APPROVED** [review]
Code-review ACK 0bf87476f55dceb106563156c7c8d6bfb8162e29

**Comment by achow101** [comment]
ACK 0bf87476f55dceb106563156c7c8d6bfb8162e29

**Pull request merged**